src/include/small_encoding.h

```diff
@@ -0,0 +1,337 @@
+  v = byte & 0x7f;
+  int shift = 7;
+  while (byte & 0x80) {
+    ::decode(byte, p);
+    v |= (T)(byte & 0x7f) << shift;
+    shift += 7;
+  }
+}
+
+// signed varint encoding
+//
+// low bit = 1 = negative, 0 = positive
+// high bit of every byte indicates whether another byte follows.
+template<typename T>
+inline void small_encode_signed_varint(T v, bufferlist& bl) {
+  uint8_t byte = 0;
+  if (v < 0) {
+    v = -v;
+    byte = 1;
+  }
+  byte |= (v & 0x3f) << 1;
+  v >>= 6;
+  while (v) {
+    byte |= 0x80;
+    ::encode(byte, bl);
+    byte = (v & 0x7f);
+    v >>= 7;
+  }
+  ::encode(byte, bl);
+}
+
+template<typename T>
+inline void small_decode_signed_varint(T& v, bufferlist::iterator& p)
+{
+  uint8_t byte;
+  ::decode(byte, p);
+  bool negative = byte & 1;
+  v = (byte & 0x7e) >> 1;
+  int shift = 6;
+  while (byte & 0x80) {
+    ::decode(byte, p);
+    v |= (T)(byte & 0x7f) << shift;
+    shift += 7;
+  }
+  if (negative) {
+    v = -v;
+  }
+}
+
+// varint + lowz encoding
+//
+// first(low) 2 bits = how many low zero bits (nibbles)
+// high bit of each byte = another byte follows
+// (so, 5 bits data in first byte, 7 bits data thereafter)
+template<typename T>
+inline void small_encode_varint_lowz(T v, bufferlist& bl) {
+  int lowz = v ? (ctz(v) / 4) : 0;
+  uint8_t byte = std::min(lowz, 3);
+  v >>= byte * 4;
+  byte |= (((uint8_t)v << 2) & 0x7c);
+  v >>= 5;
+  while (v) {
+    byte |= 0x80;
+    ::encode(byte, bl);
+    byte = (v & 0x7f);
+    v >>= 7;
+  }
+  ::encode(byte, bl);
+}
+
+template<typename T>
+inline void small_decode_varint_lowz(T& v, bufferlist::iterator& p)
+{
+  uint8_t byte;
+  ::decode(byte, p);
+  int shift = (byte & 3) * 4;
+  v = ((byte >> 2) & 0x1f) << shift;
+  shift += 5;
+  while (byte & 0x80) {
+    ::decode(byte, p);
+    v |= (T)(byte & 0x7f) << shift;
+    shift += 7;
+  }
+}
+
+// signed varint + lowz encoding
+//
+// first(low) 2 bits = how many low zero bits (nibbles)
+// high bit of each byte = another byte follows
+// (so, 5 bits data in first byte, 7 bits data thereafter)
+template<typename T>
+inline void small_encode_signed_varint_lowz(T v, bufferlist& bl) {
+  uint8_t byte = 0;
+  if (v < 0) {
+    v = -v;
+    byte = 1;
+  }
+  int lowz = v ? (ctz(v) / 4) : 0;
+  lowz = std::min(lowz, 3);
+  byte |= lowz << 1;
+  v >>= lowz * 4;
+  byte |= (((uint8_t)v << 3) & 0x78);
+  v >>= 4;
+  while (v) {
+    byte |= 0x80;
+    ::encode(byte, bl);
+    byte = (v & 0x7f);
+    v >>= 7;
+  }
+  ::encode(byte, bl);
+}
+
+template<typename T>
+inline void small_decode_signed_varint_lowz(T& v, bufferlist::iterator& p)
+{
+  uint8_t byte;
+  ::decode(byte, p);
+  bool negative = byte & 1;
+  int shift = (byte & 6) * 2;
+  v = ((byte >> 3) & 0xf) << shift;
+  shift += 4;
+  while (byte & 0x80) {
+    ::decode(byte, p);
+    v |= (T)(byte & 0x7f) << shift;
+    shift += 7;
+  }
+  if (negative) {
+    v = -v;
+  }
+}
+
+
+template<typename T>
+inline void small_delta_encode_varint_lowz(T prev, uint32_t cur,
```

Comment:
Lacks corresponding decode. No much value in this method though, can be easily implemented via small_encode_variant_lowz...